PocketKit/Sources/SharedPocketKit/Migrations/LegacyUserMigration.swift

```diff
@@ -0,0 +1,176 @@
+        }
+
+        guard let decryptedData = decryptedData else {
+            throw LegacyUserMigrationError.missingData
+        }
+
+        let legacyStore: LegacyStore
+        do {
+            legacyStore = try JSONDecoder().decode(LegacyStore.self, from: decryptedData)
+        } catch {
+            throw LegacyUserMigrationError.failedDeserialization(error)
+        }
+
+        guard isRequired(version: legacyStore.version) else {
+            updateUserDefaults()
+            return false
+        }
+
+        appSession.currentSession = Session(
+            guid: legacyStore.guid,
+            accessToken: legacyStore.accessToken,
+            userIdentifier: legacyStore.userIdentifier
+        )
+
+        updateUserDefaults()
+        return true
+    }
+
+    public func forceSkip() {
+        updateUserDefaults()
+    }
+}
+
+extension LegacyUserMigration {
+    func isRequired(version: String) -> Bool {
+        // On a fresh install, the app will not have a "previous version", and
+        // thus will not be required to run. If the previous version is < 8,
+        // and we have not yet run the migration, then it is required.
+        return required(for: version) && !previouslyRun
+    }
+
+    private var previouslyRun: Bool {
+        userDefaults.bool(forKey: Self.migrationKey)
+    }
+
+    private func required(for version: String) -> Bool {
+        guard let majorComponent = version.components(separatedBy: ".").first,
+              let previousMajorComponent = Int(majorComponent),
+              previousMajorComponent < 8 else {
+            return false
+        }
+
+        return true
+    }
+
+    private func updateUserDefaults() {
+        if !userDefaults.bool(forKey: Self.migrationKey) {
+            userDefaults.set(true, forKey: Self.migrationKey)
+        }
+    }
+
+    private var currentPassword: String? {
+        // The legacy app first checks its keychain for a password for decryption,
+        // followed by a user defaults fallback
+        return keychainPassword ?? userDefaultsPassword ?? nil
+    }
+
+    private var keychainPassword: String? {
+        LegacyPasswordRetriever(keychain: keychain).password
+    }
+
+    private var userDefaultsPassword: String? {
+        userDefaults.string(forKey: Self.decryptionKey)
+    }
+}
+
+private struct LegacyStore: Decodable {
+    let guid: String
+    let accessToken: String
+    let userIdentifier: String
+    let version: String
+
+    enum CodingKeys: String, CodingKey {
+        case guid
+        case accessToken
+        case userIdentifier = "uid"
+        case version
+    }
+}
+
+private class LegacyPasswordRetriever {
+    let password: String?
+
+    init(keychain: Keychain) {
+        let query: CFDictionary = [
+            kSecClass: kSecClassGenericPassword,
+            kSecAttrService: "Pocket",
+            kSecAttrAccount: "encryption",
+            kSecAttrAccessGroup: "group.com.ideashower.ReadItLaterProAlphaNeue",
```

Comment:
Is there a better way to do this, or do we create a ticket as a reminder to update this once we're in the true app space?

Reply:
A better way to query the keychain, or a better way to do something else?

Reply:
@nzeltzer thanks for requesting clarification. the "access group" will be different for the final release vs app store. it's basically a reflection of the bundle identifier, prefixed with `group.`

Reply:
@nzeltzer i'd have to test to see if we can retrieve the bundle id since we're technically in a swift package that isn't the app itself. let me try that out.

Reply:
it's also tricky since we potentially run the migration from the extension, which has a different bundle ID than the app, and we are required to use specific access groups.

Reply:
It looks like `Bundle.main` when called in the migration returns the correct bundle ID, so we might be able to generate this dynamically.